core/emulate.c

```diff
@@ -292,6 +292,41 @@ static const struct em_opcode_t opcode_table_0F3A[256] = {
     X16(N), X16(N), X16(N), X16(N),
 };
 
+static uint64_t gpr_read(struct em_context_t *ctxt,
+                         unsigned index, size_t size)
+{
+    uint64_t value = 0;
+
+    if (!(ctxt->gpr_cache_r & (1 << index))) {
+        ctxt->gpr_cache[index] = ctxt->ops->read_gpr(ctxt->vcpu, index);
+        ctxt->gpr_cache_r |= (1 << index);
+    }
+    memcpy(&value, &ctxt->gpr_cache[index], size);
+    return value;
+}
+
+static void gpr_write(struct em_context_t *ctxt,
+                      unsigned index, size_t size, uint64_t value)
+{
+    if (!(ctxt->gpr_cache_r & (1 << index))) {
+        ctxt->gpr_cache[index] = ctxt->ops->read_gpr(ctxt->vcpu, index);
+    }
+    ctxt->gpr_cache_r |= (1 << index);
+    ctxt->gpr_cache_w |= (1 << index);
+    memcpy(&ctxt->gpr_cache[index], &value, size);
+}
+
+static void gpr_cache_flush(struct em_context_t *ctxt)
+{
+    unsigned i;
+
+    for (i = 0; i < 16; i++) {
+        if (ctxt->gpr_cache_w & (1 << i))
+            ctxt->ops->write_gpr(ctxt->vcpu, i, ctxt->gpr_cache[i]);
+    }
+    ctxt->gpr_cache_w = 0;
```

Comment:
Because `ctxt->gpr_cache_r` isn't reset here, wouldn't that cause potential registers vs. cache mismatches, on re-entry of `em_emulate_insn` when a prior call returned `OP_WRITE_PENDING`?
Same question for `OP_READ_PENDING`?

Reply:
No, because *OP_WRITE_PENDING* and *OP_READ_PENDING* will make the corresponding `operand_read` and `operand_write` functions return with `EM_EXIT_MMIO` status, which then triggers a `goto exit`. Note that `gpr_cache_flush` is never called after the "exit" label.

Reply:
At first I also frowned upon the lack of explicit invalidation/reset of the read cache, besides the "initialization" at the beginning of `em_decode_insn()`. One possible option is to invalidate it also in `vcpu.c` right before the re-entry of `em_emulate_insn()`, in case user space (QEMU) updates any vCPU GPR behind our back. But then I scratched the idea, because QEMU's handler for `HAX_EXIT_FAST_MMIO` (`hax_handle_fastmmio()`) doesn't modify vCPU state. So invalidating the read cache only before decoding every instruction should be good enough.